gcc/rust/typecheck/rust-hir-type-check-type.cc

```diff
@@ -442,12 +442,9 @@ TypeCheckType::resolve_root_path (HIR::TypePath &path, size_t *offset,
       root_tyty = lookup;
 
       // this enforces the proper get_segments checks to take place
-      bool is_adt = root_tyty->get_kind () == TyTy::TypeKind::ADT;
-      if (is_adt)
+      if (auto adt = root_tyty->try_as<const TyTy::ADTType> ())
```

Comment:
Hmm i am not sure this is safe to do because try_as will return null if it is not an ADT Type so its not required to turn this into an ADT here.

Reply:
actually i am sorry i somehow read that as a for loop. But i would still prefer the old check, this is hard to read

Reply:
@philberty  This is something we should discuss more in depth, because you liked the API when I introduced it and now I am using it heavily in unpublished borrowchecker code.

Reply:
I like the API i just think in that specific if statement I don't like introducing variables like that. It looks too much like a for loop.

Reply:
I don't have any problem with the try_as, is and as API I think its great.

Reply:
I agree that it would be more readable to initialize the variable above the `if` and then check for its validity in the `if`'s condition